executor/runcexecutor/executor.go

```diff
@@ -326,14 +339,11 @@ func (w *runcExecutor) Run(ctx context.Context, id string, root executor.Mount,
 		return nil, err
 	}
 
-	cgroupPath := spec.Linux.CgroupsPath
-	if cgroupPath != "" {
-		return w.resmon.RecordNamespace(cgroupPath, resources.RecordOpt{
-			Release:        releaseContainer,
-			NetworkSampler: namespace,
-		})
+	if rec == nil {
+		return nil, releaseContainer(context.TODO())
 	}
-	return nil, releaseContainer(context.TODO())
+
+	return rec, rec.CloseAsync(releaseContainer)
```

Comment:
Instead of doing this, can we not spawn a goroutine here, so we can have a standard `Close` function:

```go
go func() {
  // errors are explicitly discarded in this example
  _ = rec.Close()  
  releaseContainer()
}()
return rec, nil
```

Reply:
I already did a new implementation before I realized why I did it this way.

If the daemon is closed before `releaseContainer` can be called, then it can leak resources. Monitor will take care of this and will block daemon from shutting down before all the release calls have been invoked. So if `releaseContainer` has not been called synchronously we still need to register it so that it is guaranteed to be called before shutdown. In a new goroutine this guarantee would not exist. https://github.com/moby/buildkit/commit/451e18c4ab1696dcf35cede2bd103ebfcca12e03

This would be cleaner if `Executor` itself would have a `Shutdown` function. Then in here, it could make it block until release has been called.

Reply:
We should add this context in a comment, I think it's too easy to accidentally remove in a refactor or similar.